### PR TITLE
i18n fix

### PR DESCRIPTION
**Pull request by thaerlabs**
This will solve both [#44](https://github.com/onefinestay/react-daterange-picker/issues/44) and [#116](https://github.com/onefinestay/react-daterange-picker/issues/116).

The issue is that a the time of the instantiation of WEEKDAYS and MONTHS, lang will be default `en`..

So now when calling `require('moment/locale/ar')` somewhere else, the WEEKDAYS and MONTHS will be based on the global `moment` app locale...

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-daterange-picker",
-  "version": "1.0.0",
+  "version": "1.0.1",
```
Comment (AlanFoster) [inline package.json]: Could we revert this line? This will allow us to control what goes into what release

File: src/calendar/CalendarMonth.jsx [hunk 2]
```diff
@@ -32,6 +26,12 @@ const CalendarMonth = React.createClass({
     value: CustomPropTypes.momentOrMomentRange,
   },
 
+  componentWillMount() {
+    const lang = moment().localeData();
```
Comment (AlanFoster) [inline src/calendar/CalendarMonth.jsx]: :question: Is this something we'd ever want to make prop driven in the future?

File: src/calendar/CalendarMonth.jsx [hunk 3]
```diff
@@ -107,7 +107,7 @@ const CalendarMonth = React.createClass({
     }
 
     return (
-      <option key={year} value={year}>{year}</option>
+      <option key={year} value={year}>{moment(year, 'YYYY').format('YYYY')}</option>
```
Comment (AlanFoster) [inline src/calendar/CalendarMonth.jsx]: What is the context around this change?

File: package.json [hunk 4]
```diff
@@ -25,6 +25,10 @@
     {
       "name": "Alan Foster",
       "email": "alan@alanfoster.me"
+    },
+    {
+      "name": "Thaer Abbas",
```
Comment (AlanFoster) [inline package.json]: :+1:

**Comment by AlanFoster** [comment]
Thanks for the PR!

Would we be able to add tests for this new work too? If you need any help with that, just ping us and we can help get you set up :+1:

File: src/calendar/CalendarMonth.jsx (hunk 2 above)
Comment (rolyatmax) [inline src/calendar/CalendarMonth.jsx]: I think making this prop-driven falls more in line with React's functional paradigm. I feel like it's better for the component to rely on either state passed into it (`props`) or its own internal state (`state`) and not rely on global state, like `moment().localeData()`. Also, if the moment locale changes, I don't think it would trigger an update of this component.

Thoughts?

File: package.json (hunk 1 above)
Comment (thaerlabs) [inline package.json]: Ok will do ASAP

File: src/calendar/CalendarMonth.jsx (hunk 2 above)
Comment (thaerlabs) [inline src/calendar/CalendarMonth.jsx]: In our case in my team, we have either `en` or `ar`, so from one to the other, there's a big change, and It requires the global moment object to set the locale, as the whole app requires different locale.. 

I don't mind having it prop driven, maybe it will allow more control on the component itself, but bear in mind that if I `require('moment/locale/${language}')`, also the internal state of the calendar will change.. So I believe that making it prop driven will require more work as:
- we have to isolate the moment object used in the component, initialize it with the prop passed, so it won't be affected by the global `require('moment/locale...')`

For us is fine as we just are working with all dates and hours with `moment`, so when the lang changes, all the dates and times will be automatically reformatted, and this PR is working great for us..

File: src/calendar/CalendarMonth.jsx (hunk 3 above)
Comment (thaerlabs) [inline src/calendar/CalendarMonth.jsx]: The year by default is simply a string, so to have it in arabic in the dropdown, we have to create a moment object in `YYYY` format, and format it in `YYYY`, so it'll be affected by the locale..

**Comment by thaerlabs** [comment]
@AlanFoster for sure tests have to be added, and I was thinking about adding a page in the demo to demonstrate the locale change, and I think the docs can also be improved

File: src/calendar/CalendarMonth.jsx (hunk 3 above)
Comment (AlanFoster) [inline src/calendar/CalendarMonth.jsx]: Ah, thanks for the clarification!

File: src/calendar/CalendarMonth.jsx (hunk 2 above)
Comment (rolyatmax) [inline src/calendar/CalendarMonth.jsx]: Sorry, I think I got lost there. How does `require('moment/locale/${language}')` change the internal state of the calendar and trigger a re-render?

An example of the work needed to make it take a `locale` as a prop is this old PR: https://github.com/onefinestay/react-daterange-picker/pull/56. We use a modified version of those changes for our internal dashboards, which have a language toggle that works without reloading the page (because the changing `locale` prop triggers a re-render of that part of render tree).

Comment (thaerlabs) [inline src/calendar/CalendarMonth.jsx]: @rolyatmax 
`require('moment/locale/${language}')` + `moment.locale(${language})` will change the locale of the global moment object, so whenever we use moment, it will have that locale set..

In our case we refresh the whole page as we need to load translations for messages etc.. so when bootstrapping the app we set the locale, before rendering the the component... 

I think the best solution should be merging #119 with #56.

Comment (rolyatmax) [inline src/calendar/CalendarMonth.jsx]: Ahh, I see. So changing the `locale` doesn't itself trigger a re-render? I think incorporating #56's approach is probably a good idea. You have any thoughts on this, @AlanFoster ?

Comment (thaerlabs) [inline src/calendar/CalendarMonth.jsx]: @AlanFoster @rolyatmax check the last commit.
I've added the changes of #56 + isolated the `moment` object used by `CalendarMonth`. So now `require('moment/locale/${language}')` anywhere in the app, won't change the locale of the datepicker. Locale is passed as a prop and it will load the locale, set the locale of the moment object used by the datepicker and trigger a re-render.

**Comment by thaerlabs** [comment]
@AlanFoster @rolyatmax 
I've updated the example, so now we can set **locale** as a **prop** #44 . Requiring `moment/locale/[whatever]` outside won't affect the component. The `locale` prop will take care of loading locale file, and trigger a **re-render** as required by @rolyatmax, also the dropdowns of month and year are fine #116.. So last but not least, tests should be added. I think I won't have much time during the week, help is appreciated in this context :dancers:

**Comment by xyx2011** [comment]
@AlanFoster  could this pr be merged? excited to see the new feature. :+1:  :)

File: src/calendar/CalendarMonth.jsx [hunk 5]
```diff
@@ -30,11 +24,39 @@ const CalendarMonth = React.createClass({
...
     onYearChange: React.PropTypes.func,
     value: CustomPropTypes.momentOrMomentRange,
+    locale: React.PropTypes.string,
+  },
+
+  _setupLocale(locale) {
+    this.moment = moment();
+
+    if (locale !== 'en') {
+      require(`moment/locale/${locale}`);
```
Comment (rolyatmax) [inline src/calendar/CalendarMonth.jsx]: I don't believe this will work with bundling tools like Webpack or Browserify which rely on static `require` statements to figure out what dependencies need to be included in the bundle at compile time.

Comment (rolyatmax) [inline src/calendar/CalendarMonth.jsx]: I think the user of this component will need to be responsible for requiring or registering their own `locale` definition with `moment`.

File: src/calendar/CalendarMonth.jsx [hunk 6]
```diff
@@ -30,11 +24,39 @@ const CalendarMonth = React.createClass({
...
+  componentWillMount() {
+    const { locale } = this.props;
+    this._setupLocale(locale);
+  },
+
+  componentWillReceiveProps(nextProps) {
+    const { locale } = nextProps;
+    if (locale !== this.props.locale) {
+      this._setupLocale(locale);
+    }
```
Comment (rolyatmax) [inline src/calendar/CalendarMonth.jsx]: If we compute the weekday/month names on the fly, we can simplify this and get rid of `componentWillMount`, `componentWillReceiveProps`, and `_setupLocale`.

For example: https://github.com/onefinestay/react-daterange-picker/pull/56/files#diff-78a3b23346ce557395cf4cc9eca5d165R78

File: src/calendar/CalendarMonth.jsx (hunk 5 above)
Comment (AlanFoster) [inline src/calendar/CalendarMonth.jsx]: :+1:

**Comment by AlanFoster** [comment]
Do you feel #56 is enough to implement the new locale support, or does this PR add new functionality that is missing within #56 ?

@miracle2k @rolyatmax @haerlabs As you guys have more context on i18n support i'd love your 2¢ on this

File: src/calendar/CalendarMonth.jsx (hunk 6 above)
Comment (miracle2k) [inline src/calendar/CalendarMonth.jsx]: My guess is that on-the-flow is fine performance-wise; we don't save more than a function call.

Having said that, one problem with my patch (#56) is that an error is raised if ``this.props.locale`` points to a locale that has not been loaded. This error is relatively hard to understand, and should be wrapped in a more "pretty" error message; Having a ``setupLocale`` method provides an opportunity for that.

File: src/calendar/CalendarMonth.jsx (hunk 5 above)
Comment (miracle2k) [inline src/calendar/CalendarMonth.jsx]: It's a problem with webpack/browserify. On the other hand, letting the user import the locale has issues of it's own, as I noticed running my patch in production:

``moment`` is a dependency of the datepicker. That means it is possible that we have our own copy of it in:

    ``project_dir/node_modules/react-daterange-picker/node_modules/moment``

Whether that is the case seems to depend, as far as I can tell, on:

1. The version of moment the user's project depends on, if it indeed does depend on moment (not unlikely).
2. The order of installation of the dependencies (i.e. the user project's own copy of moment is added later).
3. The version of npm (version 3 has a flattened node_modules by default).

Long story short, when ``react-daterange-picker`` executes ``require("moment")`` either one of these two paths might be imported, and in the worst case, it's not predictable which one:

    ``project_dir/node_modules/react-daterange-picker/node_modules/moment``
    ``project_dir/node_modules/moment``

To make locales work, the user would have to import the desired locale from the correct directory too.

One solution to this might be to make ``moment`` a peer dependency.

**Comment by miracle2k** [comment]
The differences between my patch in 56 and this one aren't so big. The require() issue might need to be addressed. I am +0 on the ``_setupLocale`` cache.

This PR adds a nice example thought, and it seems @thaerlabs has more interesting in moving this forward.

File: src/calendar/CalendarMonth.jsx (hunk 5 above)
Comment (rolyatmax) [inline src/calendar/CalendarMonth.jsx]: :+1: to making `moment` a peer dependency at some point

**Comment by AlanFoster** [comment]
Great, thanks for the input @miracle2k :+1:

Getting i18n support would obviously be great to have merged and released; are you happy to do the remaining work for this PR @thaerlabs to help get this PR shipped?

**Comment by miracle2k** [comment]
Can we get this merged? Seems @thaerlabs  made some changes.

**Comment by AlanFoster** [comment]
@thaerlabs Looks good to me; Would you mind adding some quick tests then we can get this merged in? :)

File: dist/calendar/CalendarMonth.js [hunk 7]
```diff
@@ -65,7 +60,33 @@ var CalendarMonth = _react2['default'].createClass({
...
+  },
+
+  _setupLocale: function _setupLocale(locale) {
+    this.moment = (0, _moment2['default'])();
+
+    this.moment.locale(locale);
+
+    var lang = this.moment.localeData(locale);
+
+    this.WEEKDAYS = _immutable2['default'].List(lang._weekdays).zip(_immutable2['default'].List(lang._weekdaysShort));
```
Comment (ashmigelski) [inline dist/calendar/CalendarMonth.js]: Hey
Its noot good to get just `lang._weekdays` or `lang._month`, its not always an array (e.g for `ru` locale https://github.com/moment/moment/blob/develop/locale/ru.js#L38)
Check https://github.com/moment/moment/blob/602718d67f06011cf6483387ad96989528f55597/src/lib/units/month.js#L65

Comment (thaerlabs) [inline dist/calendar/CalendarMonth.js]: @ashmigelski actually the problem is that there's an issue with the locale definition in the `ru` locale. When locale is `ru`, check `lang.hasOwnProperty('_months')`, for russian it will return `false`, which means that the months are not in the locale data so it's being loaded from the defaults in the prototype (english). That's why in the dropdown of the months there are russian months

**Comment by thaerlabs** [comment]
@AlanFoster I'll work on the tests ASAP. I think also that we should enhance the documentation, especially for the fact that this feature requires the user to `require` the `moment/locale` manually, otherwise setting the prop won't be enough to change the datepicker. This approach is better than dynamic `require` the locale in the `_setupLocale` function when `componentWillReceiveProps`, because when webpack bundles, it will put there all the `moment/locale/**`, as it doesn't know which one to pic.

**Comment by thaerlabs** [comment]
@AlanFoster can you help me in testing the months dropdown and the weekdays text? 
I want to assert against the text content of the months and weekdays when changing locale

File: dist/calendar/CalendarMonth.js (hunk 7 above)
Comment (ashmigelski) [inline dist/calendar/CalendarMonth.js]: @thaerlabs actually its not `ru` locale problem. Moment supports 2 way month localization:
1) array of month names, like `en`
```
_months: ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
```
2) object with months names in several forms, like `ru`, `ka`, `be`, `lt` and [others](https://github.com/moment/moment/search?utf8=%E2%9C%93&q=standalone)
```
_months : {
   format: 'января_февраля_марта_апреля_мая_июня_июля_августа_сентября_октября_ноября_декабря'.split('_'),
   standalone: 'январь_февраль_март_апрель_май_июнь_июль_август_сентябрь_октябрь_ноябрь_декабрь'.split('_')
}
```
Here how it works [inside Moment](https://github.com/moment/moment/blob/602718d67f06011cf6483387ad96989528f55597/src/lib/units/month.js#L65) 
```
export function localeMonths (m, format) {
    return isArray(this._months) ? this._months[m.month()] :
        this._months[MONTHS_IN_FORMAT.test(format) ? 'format' : 'standalone'][m.month()];
}
```
Just use same logic in your code

**Comment by fluse** [comment]
when will this merged?

**Comment by evgenTraytyak** [comment]
@AlanFoster ping

File: example/index.jsx [hunk 8]
```diff
@@ -91,13 +91,30 @@ const DatePickerSingle = React.createClass({
...
+      locale: 'en',
+    };
+  },
+
   getDefaultProps() {
     return {};
   },
 
+  _selectLocale() {
+    require(`moment/locale/${this.refs.locale.value}`);
```
Comment (AlanFoster) [inline example/index.jsx]: Do we need this line, if we've already loaded it as part of `example/code-snippets/i18n.jsx` ?

File: src/calendar/CalendarMonth.jsx [hunk 9]
```diff
@@ -30,11 +24,31 @@ const CalendarMonth = React.createClass({
     onMonthChange: React.PropTypes.func,
     onYearChange: React.PropTypes.func,
     value: CustomPropTypes.momentOrMomentRange,
+    locale: React.PropTypes.string,
+  },
+
+  _setupLocale(locale) {
```
Comment (AlanFoster) [inline src/calendar/CalendarMonth.jsx]: can we remove the underscore prefix here?

File: example/code-snippets/i18n.jsx [hunk 10]
```diff
@@ -0,0 +1,68 @@
+import React from 'react';
+
+import DateRangePicker from 'react-daterange-picker';
+import moment from 'moment';
+
+/**
+ * Load moment locales that you need in your app
+ * to make the localeData available to the DatePicker
+ */
+require('moment/locale/en');
```
Comment (AlanFoster) [inline example/code-snippets/i18n.jsx]: To make the date picker usable, should we require `moment/locale/en` by default, and set the default proptypes for locale to be `en` - that we we're not making such a backwards compatible change for existing users?

File: src/calendar/tests/CalendarMonth.spec.js [hunk 11]
```diff
@@ -35,6 +35,7 @@ describe('The CalendarMonth Component', function () {
         onMonthChange: function () {},
         onYearChange: function () {},
         bemBlock: 'DateRangePicker',
+        locale: 'en',
```
Comment (AlanFoster) [inline src/calendar/tests/CalendarMonth.spec.js]: We could provide default proptypes and leave this as-s?

File: example/code-snippets/i18n.jsx (hunk 10 above)
Comment (AlanFoster) [inline example/code-snippets/i18n.jsx]: i.e. `react-daterange-picker` by default loads in `require('moment/locale/en')`, so by default it is locale aware - and if you wish to change it in your application it's easier to do so

**Comment by AlanFoster** [comment]
@thaerlabs Would love to get this out; What testing difficulties did you have? :+1:

**Comment by evgenTraytyak** [comment]
@thaerlabs russian locale `ru` doesn't work :(
`Uncaught TypeError: Expected Array or iterable object of values: [object Object]`

**Comment by onzag** [comment]
@thaerlabs is this merged? how can I check this? do you have a bit of documentation? I need to use this in production :+1: 

Sorry I'm a bit of a newbie with git I can't find the repo.

**Comment by maullerz** [comment]
> Uncaught TypeError: Expected Array or iterable object of values: [object Object]
> belarusian locale be too..

its because in moment@2.11.0 they change format of weekdays and months for some locales
[RU https://github.com/moment/moment/blob/develop/locale/ru.js](https://github.com/moment/moment/blob/develop/locale/ru.js)
so its now objects, not arrays

i didn`t found any issue or PR about that, only this [commit](https://github.com/moment/moment/commit/8016bbe6c874e6af5fe61bc15a614058454c8396)

**Comment by fluse** [comment]
@AlanFoster please merge branches!!!!!!!

**Comment by thaerlabs** [comment]
@AlanFoster I added tests for the month and year headers. Please check it and provide your feedback!

**Comment by thaerlabs** [comment]
@evgenTraytyak @maullerz the problem is with the locale definition for both `ru` and `be`. Somebody should fix the format and open a pull request to `moment`. I don't think it's a good idea to work around this bug in moment in the datepicker component, as if they fix it, it'll break again the datepicker. Plus we should add a special case for those locales with a different format, which will make the component more fragile. that's my thought..

**Comment by maullerz** [comment]
agree with @thaerlabs
this `ru` and `be` issues should be fixed in moment.js, not in datepicker
@AlanFoster please merge branches!

**Comment by ashmigelski** [comment]
Guys, you cant merge with `ru` locale issue. And more over there's no bug in `moment.js`...

Here's my workaround, just add `isArray` check for ` lang._weekdays` and `lang._months` in https://github.com/onefinestay/react-daterange-picker/blob/master/src/calendar/CalendarMonth.jsx#L12

```js
    const lang = moment().localeData();

    let weekdays = lang._weekdays;
    if (!Array.isArray(weekdays)) {
        weekdays = weekdays.standalone;
    }
    let months = lang._months;
    if (!Array.isArray(months)) {
        months = months.standalone;
    }

    const WEEKDAYS = Immutable.List(weekdays).zip(Immutable.List(lang._weekdaysShort));
    const MONTHS = Immutable.List(months);
```

**Comment by maullerz** [comment]
> Guys, you cant merge with ru locale issue. And more over there's no bug in moment.js...

why? can you explain?

**Comment by ashmigelski** [comment]
> why? can you explain?

I explained it 3 month ago https://github.com/onefinestay/react-daterange-picker/pull/119#discussion_r56464224

In short
- datepicker expect that `_monthNames` can be only an array. 
- `moment` support different locales and some of them have month names in several forms.
- for such locales `_monthNames` becomes an object with arrays inside

If you want  datepicker work well with **all** locations, you have to add check is locale is simple (`_monthNames` just array) or complex (`_monthNames` is an object and real month names can be found inside, like `_monthNames['standalone']` ) 

Code in my prev comment

**Comment by wmertens** [comment]
Moment has an API to get the months and weekdays as an array. Call that instead of relying on Moment internals: http://momentjs.com/docs/#/i18n/listing-months-weekdays/

**Comment by thaerlabs** [comment]
@wmertens great hint! I'll fix that ASAP

**Comment by AlanFoster** [comment]
@thaerlabs Let me know when you've had the time to apply the PR feedback from @wmertens
After a squash of commits I'll try to get this in! :+1:

File: src/calendar/tests/CalendarMonth.spec.js (hunk 11 above)
Comment (AlanFoster) [inline src/calendar/tests/CalendarMonth.spec.js]: @thaerlabs Would you mind adding this default functionality as part of this PR?

It means that this PR could go out without breaking API backwards compatability

File: src/tests/i18n.spec.js [hunk 12]
```diff
@@ -0,0 +1,133 @@
+import React from 'react';
+import TestUtils from 'react-addons-test-utils';
+import moment from 'moment';
+import _ from 'underscore';
+import CalendarMonth from '../calendar/CalendarMonth';
+import CalendarDate from '../calendar/CalendarDate';
+
+describe('Localization', function () {
```
Comment (AlanFoster) [inline src/tests/i18n.spec.js]: 💯  For adding tests

File: src/tests/i18n.spec.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+  const testLocale = 'ar';
```
Comment (AlanFoster) [inline src/tests/i18n.spec.js]: @thaerlabs Would we want to add a test that switches this testLocale ?

File: src/tests/i18n.spec.js [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+    require(`moment/locale/${testLocale}`);
+    moment.locale(testLocale);
+    this.useShallowRenderer({
+      locale: testLocale,
+    });
+
+    const headerYearSelect = this.container.props.children[2].props.children[1];
+
+    _.map(_.compact(headerYearSelect.props.children), (option, index) => {
```
Comment (AlanFoster) [inline src/tests/i18n.spec.js]: @thaerlabs Could we change this to `each` - as we simply want to loop over every value for test purposes

Also, should we add an assertion here that there are actually any children present?

**Comment by AlanFoster** [comment]
@thaerlabs Requires a rebase :)
Let me know if I can do anything to help out :+1:

**Comment by maullerz** [comment]
@thaerlabs any news about that? :pray:

**Comment by andywer** [comment]
@AlanFoster @thaerlabs Is there anything left to do? We also stumbled upon the localization issue here.

**Comment by maullerz** [comment]
@andywer fyi
there is rebased and fixed PR forked from this - #150 
so i use react-daterange-picker from this branch - git://github.com/maullerz/react-daterange-picker.git#i18n-fix-build
Temp workaround while we all waiting updates of original repo.

**Comment by AlanFoster** [comment]
Closing this in favour of #150

Thanks for the hard work guys; I'll aim to release this as 2.x
